datadog/resource_datadog_cloud_workload_security_agent_rule.go

```diff
@@ -0,0 +1,168 @@
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV2 := providerConf.DatadogClientV2
+	authV2 := providerConf.AuthV2
+
+	agentRuleCreate := buildCwsAgentRuleCreatePayload(d)
+
+	response, httpResponse, err := datadogClientV2.CloudWorkloadSecurityApi.CreateCloudWorkloadSecurityAgentRule(authV2, *agentRuleCreate)
+	if err != nil {
+		return utils.TranslateClientErrorDiag(err, httpResponse, "error creating cloud workload security agent rule")
+	}
+	if err := utils.CheckForUnparsed(response); err != nil {
+		return diag.FromErr(err)
+	}
+
+	// update the resource
+	updateResourceDataAgentRuleFromResponse(d, response)
+
+	return nil
+}
+
+func resourceDatadogCloudWorkloadSecurityAgentRuleRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV2 := providerConf.DatadogClientV2
+	authV2 := providerConf.AuthV2
+
+	id := d.Id()
+	agentRuleResponse, httpResponse, err := datadogClientV2.CloudWorkloadSecurityApi.GetCloudWorkloadSecurityAgentRule(authV2, id)
+	if err != nil {
+		if httpResponse != nil && httpResponse.StatusCode == 404 {
+			d.SetId("")
+			return nil
+		}
+		return utils.TranslateClientErrorDiag(err, httpResponse, "error fetching cloud workload security agent rule")
+	}
+	if err := utils.CheckForUnparsed(agentRuleResponse); err != nil {
+		return diag.FromErr(err)
+	}
+
+	updateResourceDataAgentRuleFromResponse(d, agentRuleResponse)
+
+	return nil
+}
+
+func resourceDatadogCloudWorkloadSecurityAgentRuleUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV2 := providerConf.DatadogClientV2
+	authV2 := providerConf.AuthV2
+
+	agentRuleId := d.Id()
+
+	agentRuleUpdate := buildCwsAgentRuleUpdatePayload(d)
+
+	if _, httpResponse, err := datadogClientV2.CloudWorkloadSecurityApi.UpdateCloudWorkloadSecurityAgentRule(authV2, agentRuleId, *agentRuleUpdate); err != nil {
+		return utils.TranslateClientErrorDiag(err, httpResponse, "error updating cloud workload security agent rule")
+	}
+
+	return nil
+}
+
+func resourceDatadogCloudWorkloadSecurityAgentRuleDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV2 := providerConf.DatadogClientV2
+	authV2 := providerConf.AuthV2
+
+	agentRuleId := d.Id()
+
+	if httpResponse, err := datadogClientV2.CloudWorkloadSecurityApi.DeleteCloudWorkloadSecurityAgentRule(authV2, agentRuleId); err != nil {
+		return utils.TranslateClientErrorDiag(err, httpResponse, "error deleting cloud workload security agent rule")
+	}
+
+	return nil
+}
+
+func updateResourceDataAgentRuleFromResponse(d *schema.ResourceData, agentRuleResponse datadogV2.CloudWorkloadSecurityAgentRuleResponse) {
+	data := agentRuleResponse.GetData()
+	d.SetId(data.GetId())
```

Comment:
```suggestion
```
Typically terraform resources set the ID in the create method.

Reply:
This is needed when we update the resource from the response

Reply:
Just tried, and I get an unknown provider error during the tests if this line is removed.

Reply:
Right because you use the same method for update and create.